sw/device/lib/dif/dif_padctrl.c

```diff
@@ -0,0 +1,180 @@
+    return kDifPadctrlEnableErrorInvalidAttr;
+  }
+
+  ptrdiff_t pad_base_register_offset;
+  uint32_t pad_register_index;
+  uint32_t pad_attr_field_index;
+
+  // Do pad index range check and calculate which register and attribute field
+  // within it needs to be modified based on the kind of pad.
+  if (kind == kDifPadctrlPadMio) {
+    if (index > (PADCTRL_PARAM_NMIOPADS - 1)) {
+      return kDifPadctrlEnableErrorInvalidPad;
+    }
+
+    pad_base_register_offset = PADCTRL_MIO_PADS0_REG_OFFSET;
+    pad_register_index = index / PADCTRL_MIO_PADS_ATTR_FIELDS_PER_REG;
+    pad_attr_field_index = index % PADCTRL_MIO_PADS_ATTR_FIELDS_PER_REG;
+  } else {
+    if (index > (PADCTRL_PARAM_NDIOPADS - 1)) {
+      return kDifPadctrlEnableErrorInvalidPad;
+    }
+
+    pad_base_register_offset = PADCTRL_DIO_PADS_REG_OFFSET;
+    pad_register_index = index / PADCTRL_DIO_PADS_ATTR_FIELDS_PER_REG;
+    pad_attr_field_index = index % PADCTRL_DIO_PADS_ATTR_FIELDS_PER_REG;
+  }
+
+  // Determine which bit the attribute field that is being modified starts at
+  // and which bit within the register controls the attribute being modified.
+  uint32_t attr_field_start_bit = pad_attr_field_index * PADCTRL_PARAM_ATTRDW;
+  uint32_t attr_bit_index = attr_field_start_bit + attr;
+
+  // Calculate register offset for the register containing the attribute field
+  // that is being modified.
+  ptrdiff_t pad_register_offset =
+      pad_base_register_offset +
+      (PADCTRL_REGISTER_WIDTH_IN_BYTES * pad_register_index);
+
+  if (enable) {
+    // Extract the attribute bits from the given pad register and attribute
+    // field
+    uint32_t attr_field_mask = PADCTRL_ATTR_FIELD_MASK << attr_field_start_bit;
+
+    uint32_t attr_reg_value =
+        mmio_region_read32(padctrl->base_addr, pad_register_offset);
+
+    uint32_t attr_field_value =
+        (attr_reg_value & attr_field_mask) >> attr_field_start_bit;
+
+    // Update the read attribute field with the attribute to enable
+    attr_field_value |= (1 << attr);
+
+    // Check if the new setting causes a conflict
+    if (dif_padctrl_conflicting_attributes(attr_field_value)) {
+      return kDifPadctrlEnableErrorConflict;
+    }
+
+    // Update the register value with the new field setting and write it back
+    // to padctrl
+    attr_reg_value = (attr_reg_value & ~attr_field_mask) |
+                     (attr_field_value << attr_field_start_bit);
+
+    mmio_region_write32(padctrl->base_addr, pad_register_offset,
+                        attr_reg_value);
+
+    // See if attribute enable was successful
+    bool test_attr = mmio_region_get_bit32(padctrl->base_addr,
+                                           pad_register_offset, attr_bit_index);
+
+    // Attribute fields are WARL so if attribute isn't supported read value
+    // won't have it set.
+    if (test_attr) {
+      return kDifPadctrlEnableErrorNone;
+    } else {
+      return kDifPadctrlEnableErrorNotSupported;
```

Comment:
What if this function is called after padctrl is locked?

Reply:
The result will then depend upon whether the attribute was enabled or not, which is a bit messy and confusing. I think this adds weight to having the the register lock check as part of this function.